### PR TITLE
prototype: add ignition phase2 for comparison with ansible

### PR DESCRIPTION
**Pull request by mikedanese**
It's less than 30mb  and I think we could get it down to 20mb, much faster, and no python dependency on host or master. Requires a slightly modified igntion https://github.com/mikedanese/ignition/tree/vanilla

If I preinstall docker it takes 2.5 minutes to bring up a cluster.

ref #160

### CODE REVIEW

File: phase2/ignition/vanilla/master.jsonnet [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+          filesystem: "root",
+          path: "/usr/local/bin/kubectl",
+          contents: {
+            source: "https://storage.googleapis.com/kubernetes-release/release/v1.3.0-beta2/bin/linux/amd64/kubectl",
+          },
+          mode: 511,
+        },
+      ],
+    },
+  }
```
Comment (mikedanese) [inline phase2/ignition/vanilla/master.jsonnet]: over usage of jsonnet may be making this harder to understand.

**Comment by mikedanese** [comment]
@roberthbailey this is a functioning node bootstrap phase if you are interested

**Comment by colemickens** [comment]
"pre-install docker" = use a VM image that already has it baked in?

**Comment by mikedanese** [comment]
@colemickens yup.

**Comment by errordeveloper** [comment]
Didn't know that ignition is not CoreOS-specific, unlike it's predecessor...

**Comment by colemickens** [comment]
(CoreOS has their own impl of coreos-cloudinit, but Ubuntu/RedHat/Arch/etc use cloudinit as well. I use it in the other azure bring-up. It's very fast since it's baked into the Ubuntu cloud images.)

Anyway, I like the looks of this so far, definitely interested in getting rid of the 300MB container. I didn't think the jsonnet was too complicated either.

Is Ignition meant to be used as a change management system? With cloudinit it seemed like it was always used for initial provisioning and I didn't observe many people using it after initial deployment for updates (IIRC it's not even straightforward to re-execute it).

Given an assumption that we wanted to support upgrades, is that something Ignition can be used for?

**Comment by errordeveloper** [comment]
I've used the CoreOS cloud-init as well as the one Ubuntu ships. Both work
okay, but I would love to see some convergence, perhaps something new
should emerge, and it could be more testable and way it integrates with
systemd could be stronger (well, may be that falls into testability). What
I found with CoreOS version is that you can literally do a lot of things
via systemd units, but the whole structure is very fragile and has no
tooling for verification. Also, systmed units aren't very expressive and
the whole syntax is a mess, you end up with a soup of YAML, systemd/INI and
some bash, all in one file. Also you have some built-in magic for etcd and
other thigs that CoreOS ships, and you soon find out that the scope is not
defined very clearly. With the plain Ubuntu version there is less - no
systemd bits and only some barely documented bits exist for installing apt
packages etc, given that it's written in Python and uses YAML, it could
just be Ansible.

I think it may be of interest to CNCF to standardise on cloud-init v2.
Mike, this combination of Jsonnet and Ignition may well be a starting point
for such a tool. I like Ignition, it has a much clearer scope and is not
meant for human consumption, especially given how you have inline your
systmed units or any bash helpers etc.

**Comment by errordeveloper** [comment]
Cole, to your point regarding updates, I am not quite sure how Ignition
could do that. IIRC, the original motivation CoreOS team had for Ignition
was that cloud-init ran over and over on reboots, and the behaviour was not
always what user expected, so Ignition was very explicitly designed to run
only once at the initial boot.

File: phase1/gce/configure-vm.sh [hunk 2]
```diff
@@ -13,12 +13,6 @@ get_metadata() {
 
 ROLE=$(get_metadata "k8s-role")
 
-mkdir -p /etc/systemd/system/docker.service.d/
-cat <<EOF > /etc/systemd/system/docker.service.d/clear_mount_propagtion_flags.conf
-[Service]
-MountFlags=shared
```
Comment (colemickens) [inline phase1/gce/configure-vm.sh]: Is this not needed anymore?

**Comment by colemickens** [comment]
If we don't want to use Ansible/Ignition to do upgrades (maybe instead we'd use the options being discussed for self-hosting? pod kubelet api, etc?)... then I think this approach makes sense.

File: phase1/gce/configure-vm.sh (hunk 2 above)
Comment (mikedanese) [inline phase1/gce/configure-vm.sh]: Added back further down in the file

**Pull request merged**